src/irmin-pack/s.ml

```diff
@@ -14,95 +14,49 @@
  * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
  *)
 
-module type ATOMIC_WRITE_STORE = sig
-  include Irmin.ATOMIC_WRITE_STORE
+type integrity_error = [ `Wrong_hash | `Absent_value ]
 
-  val v : ?fresh:bool -> ?readonly:bool -> string -> t Lwt.t
+module type CLOSEABLE = sig
+  type 'a t
 
-  val flush : t -> unit
+  val close : _ t -> unit Lwt.t
 end
 
-module type LAYERED_CONTENT_ADDRESSABLE_STORE = sig
-  include Pack.S
-
-  module U : Pack.S with type value = value
-
-  module L : Pack.S
-
-  val v :
-    [ `Read ] U.t ->
-    [ `Read ] U.t ->
-    [ `Read ] L.t option ->
-    flip:bool ->
-    freeze_lock:Lwt_mutex.t ->
-    add_lock:Lwt_mutex.t ->
-    [ `Read ] t
-
-  type layer_id = [ `Upper1 | `Upper0 | `Lower ]
-
-  val layer_id : [ `Read ] t -> key -> layer_id Lwt.t
-
-  type 'a layer_type =
-    | Upper : [ `Read ] U.t layer_type
-    | Lower : [ `Read ] L.t layer_type
-
-  val copy :
-    'l layer_type * 'l ->
-    [ `Read ] t ->
-    ?aux:(value -> unit Lwt.t) ->
-    string ->
-    key ->
-    unit Lwt.t
-
-  val check_and_copy :
-    'l layer_type * 'l ->
-    [ `Read ] t ->
-    ?aux:(value -> unit Lwt.t) ->
-    string ->
-    key ->
-    unit Lwt.t
+module type CHECKABLE = sig
+  type 'a t
 
-  val mem_lower : 'a t -> key -> bool Lwt.t
-
-  val mem_next : [> `Read ] t -> key -> bool Lwt.t
-
-  val flip_upper : 'a t -> unit
-
-  val current_upper : 'a t -> [ `Read ] U.t
+  type key
 
-  val next_upper : 'a t -> [ `Read ] U.t
+  val integrity_check :
+    offset:int64 -> length:int -> key -> _ t -> (unit, integrity_error) result
+end
 
-  val lower : 'a t -> [ `Read ] L.t
+module type LAYERED_GENERAL = sig
+  type 'cap t
```

Comment:
I prefer `a t` here instead of `cap`, if it's ok with you.

Reply:
Sure :+1: